### PR TITLE
Enemy Spawner

### PR DESCRIPTION
**Pull request by rob1997**
Enemy Spawner Implemented with Object Pooling and Spawn Points

### CODE REVIEW

File: Assets/Scripts/NPC/Enemies/EnemySpawner.cs [hunk 1]
```diff
@@ -12,15 +15,62 @@ public class EnemySpawner : MonoBehaviour
     {
         #region Fields
 
-        [SerializeField] private GameObject[] enemies;
+        [SerializeField] private EnemyController enemyPrefab;
```
Comment (kantagara) [inline Assets/Scripts/NPC/Enemies/EnemySpawner.cs:18]: I'd make this an array. We'll maybe have more than one enemies.

File: Assets/Scripts/NPC/Enemies/EnemySpawner.cs [hunk 2]
```diff
@@ -12,15 +15,62 @@ public class EnemySpawner : MonoBehaviour
...
+        
+        [SerializeField] private Transform[] spawnPoints;
+
+        private ObjectPool<EnemyController> _pool;
```
Comment (kantagara) [inline Assets/Scripts/NPC/Enemies/EnemySpawner.cs:22]: OMG UNITYS OBJECT POOL!!!

**Review by kantagara — APPROVED** [review]
Aweomse!

File: Assets/Scripts/NPC/Enemies/EnemySpawner.cs (hunk 1 above)
Comment (kantagara) [reply Assets/Scripts/NPC/Enemies/EnemySpawner.cs:18]: Ahh but yeah makes sense that each enemy type has it's own spawner, nvm.

**Review by sneakzttv — APPROVED** [review]
yeet object pooling, amazing, never seen it in use before, great way to optimize https://learn.unity.com/tutorial/introduction-to-object-pooling#5ff8d015edbc2a002063971c

File: Assets/Scripts/NPC/Enemies/EnemySpawner.cs (hunk 2 above)
Comment (oleksandrchainsafe) [reply Assets/Scripts/NPC/Enemies/EnemySpawner.cs:22]: 😱🤯

File: Assets/Scripts/NPC/NPCController.cs [hunk 3]
```diff
@@ -6,26 +6,23 @@
 
 namespace Chainsafe.BlockBlasterz
 {
-    [RequireComponent(typeof(NavMeshAgent), typeof(Vision))]
-    public class NPCController : MonoBehaviour
+    [RequireComponent(typeof(NavMeshAgent))]
+    public class NPCController : Damageable.Damageable
```
Comment (oleksandrchainsafe) [inline Assets/Scripts/NPC/NPCController.cs:10]: NpcController, because NPC is more than 2 symbols.

**Review by oleksandrchainsafe — CHANGES_REQUESTED** [review]
All looks good 👍
Built-in pool - very cool 🔥

The only thing I don't like is Damageable->NpcController->EnemyController mega-hierarchy.

Healthy way would be to split Enemy into a set of components. Each component is responsible for it's own thing (health for example handles damage and sends 'Dead' event when health <= 0). The composition of the components makes gameobject into an enemy, that can walk, take damage, shoot player and so on. Another enemy can do all of that + explode on death because it has one more component that makes him spawn explosion on death.

File: Assets/Scripts/NPC/NPCController.cs (hunk 3 above)
Comment (rob1997) [reply Assets/Scripts/NPC/NPCController.cs]: Yep makes sense, I've made the changes

**Review by oleksandrchainsafe — APPROVED** [review]

**Pull request merged**